project/ian-chiu/README.md

```diff
@@ -0,0 +1,185 @@
+
+#### State
+
+The maximum number of state a cell can have. If a cell's state is greater this number, then it died (back to state 0).
+
+#### Neighbor
+
+There are two kinds of neighbor, Von Neumann(VN) and Moore(M). Von Neumann is like a rubik cube, the cell at the center has 26 neighbors. Moore has only 4 neighbors, top, down, front and back.
+
+### Optimization
+
+#### Multithreading and SIMD... or any parallel computing techinques
+
+Use parallel computing techinques, like multithreading and SIMD operations, to increase computation speed of calculating all cell states in the next frame under a rule. 
+
+If a grid is 100x100x100, we have to calculate next states of 1000000 cells according to their neighbors every frame, which needs a lot of computation.
+
+#### Greedy Meshing Algorithm
+
+The idea of the algorithm is combining adjacent blocks to reduce the part count while visually keeping everything the same, which is useful to reduce draw calls. 
+
+For a 100x100x100 grid, we have to draw maximum 1000000 cubes if all cells are not empty. If the number of draw calls is too large, the communication between cpu and gpu will be overwhelming, which can greatly reduce the performance of rendering.
+
+![](https://i.imgur.com/SCIFdNj.png)
+
+## Prospective Users
+
+### Biology
+
+Several biological processes and systems are simulated using Celluar automations, including the patterns of some seashells, moving wave patterns on the skin of cephalopods and the behaviors of brain neurons.
+
+### Physics
+
+Celluar automata are used to simulate and study physical phenomena, such as gas and fluid dynamics.
+
+### Cryptography
+
+Celluar automata are proposed for use in public key cryptography. They can be utilized to construct pseudorandom number generators, and to design error correction codes.
+
+### Generative Art and Music
+
+Cellular automata have been used in generative music and evolutionary music composition and procedural terrain generation in video games.
+
+*Sources*:
+* Tech Target: https://www.techtarget.com/searchenterprisedesktop/definition/cellular-automaton
+* Wikipedia:
+https://en.wikipedia.org/wiki/Cellular_automaton#Applications
+
+## System Architecture
+
+### Features
+
+The software should have following features:
+
+1. Simulation operations: randomize, run, pause, clear
+2. Simulation settings: randomize density and range, grid size
+3. Some cell rules can be used by default
+4. Users can add their own cells rules
+5. Basic camera control: Orbiting and Zooming
+
+### Input / Output
+
+* Input: GUI, keyboard, mouse
+* Ouput: application window
+
+### Third Party Library
+
+- raylib-py or pygame: A simple video games programming library for dealing with windows, keyboard/mouse input, graphics and GUI.
+
+## API Description
+
+### Python
+
+* Rule: 
+    * apply: apply the cell rule
```

Comment:
Would you need to call back to Python for enforcing the rules, or the rules are only data to be set in the C++ engine?

Reply:
The rule should be written in python, and C++ will get the rule from python to calculate the next state of all cells. However, I am not 100% sure. I will make the decision during development.

Reply:
Sounds good.